File: leetcode/medium/102-level-order/recursive.cpp
/**
 * @file q.cpp
 * @author Vortexx2
 * @brief Problem 102 - Binary Tree Level Order Traversal
 * @date 24-07-2021
 * 
 * Runtime - 0 ms
 * Memory Usage - 13.8 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  vector<vector<int>> levelOrder(TreeNode *root)
  {
    vector<vector<int>> res;
    if (!root)
      return res;

    helper(root, res, 0);

    return res;
  }

  void helper(TreeNode *curr, vector<vector<int>> &arr, int level)
  {

    if (!curr)
      return;

    if (level >= arr.size())
    {
      vector<int> lev;
      lev.push_back(curr->val);
      arr.push_back(lev);
    }

    else
      arr[level].push_back(curr->val);

    helper(curr->left, arr, level + 1);
    helper(curr->right, arr, level + 1);
  }
};

File: leetcode/medium/102-level-order/recursive.py
"""
  @author Vortexx2
  Problem 102 - Binary Tree Level Order Traversal

  Runtime - 36 ms
  Memory Usage - 15.1 MB
"""
from typing import List


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def levelOrder(self, root: TreeNode) -> List[List[int]]:
    res = []

    if not root:
      return res

    self.traverse(root, res, 0)
    return res

  def traverse(self, curr: TreeNode, arr: List[List[int]], level: int) -> None:

    if not curr:
      return

    if level >= len(arr):
      arr.append([curr.val])

    else:
      arr[level].append(curr.val)

    self.traverse(curr.left, arr, level + 1)
    self.traverse(curr.right, arr, level + 1)
